core/finality-grandpa/src/lib.rs

```diff
@@ -537,232 +758,246 @@ impl<B, E, Block: BlockT<Hash=H256>, N, RA> voter::Environment<Block::Hash> for
 	fn precommit_equivocation(
 		&self,
 		_round: u64,
-		equivocation: Equivocation<Self::Id, Precommit<Block::Hash>, Self::Signature>
+		equivocation: Equivocation<Self::Id, Precommit<Block>, Self::Signature>
 	) {
 		warn!(target: "afg", "Detected precommit equivocation in the finality worker: {:?}", equivocation);
 		// nothing yet
 	}
 }
 
-/// Run a GRANDPA voter as a task. The returned future should be executed in a tokio runtime.
-pub fn run_grandpa<B, E: Send + Sync, Block: BlockT<Hash=H256>, N, RA: Send + Sync + 'static>(
-	config: Config,
-	client: Arc<Client<B, E, Block, RA>>,
-	voters: HashMap<AuthorityId, usize>,
-	network: N,
-) -> Result<impl Future<Item=(),Error=()>,client::error::Error> where
-	Block::Hash: Ord,
-	B: Backend<Block, Blake2Hasher> + 'static,
-	E: CallExecutor<Block, Blake2Hasher> + 'static,
-	N: Network + 'static,
-	N::In: 'static,
-	NumberFor<Block>: As<u32>,
-{
-	let chain_info = client.info()?;
-	let genesis_hash = chain_info.chain.genesis_hash;
-	let last_finalized = (
-		chain_info.chain.finalized_hash,
-		chain_info.chain.finalized_number.as_()
-	);
+/// Client side of the GRANDPA APIs declared in fg-primitives.
+pub trait ApiClient<Block: BlockT<Hash=H256>> {
+	/// Get the genesis authorities for GRANDPA.
+	fn genesis_authorities(&self) -> Result<Vec<(AuthorityId, u64)>, ClientError>;
 
-	let (last_round_number, last_state) = match client.backend().get_aux(LAST_COMPLETED_KEY)? {
-		None => (0, RoundState::genesis((genesis_hash, 0))),
-		Some(raw) => <(u64, RoundState<Block::Hash>)>::decode(&mut &raw[..])
-			.ok_or_else(|| ::client::error::ErrorKind::Backend(
-				format!("Last GRANDPA round state kept in invalid format")
-			))?
-	};
+	/// Check a header's digest for a scheduled change.
+	fn scheduled_change(&self, header: &Block::Header)
+		-> Result<Option<ScheduledChange<NumberFor<Block>>>, ClientError>;
+}
 
-	let environment = Arc::new(Environment {
-		inner: client,
-		config,
-		voters,
-		network,
-	});
+impl<B, E, Block: BlockT<Hash=H256>, RA> ApiClient<Block> for Arc<Client<B, E, Block, RA>> where
+	B: Backend<Block, Blake2Hasher> + 'static,
+	E: CallExecutor<Block, Blake2Hasher> + 'static + Clone + Send + Sync,
+	DigestFor<Block>: Encode,
+	RA: Send + Sync,
+{
+	fn genesis_authorities(&self) -> Result<Vec<(AuthorityId, u64)>, ClientError> {
+		use runtime_primitives::traits::Zero;
+
+		self.call_api_at_strong(
```

Comment:
This should be changed to `self.runtime_api().grandpa_authorities(&BlockId::Number(NumberFor::<Block>::zero())`.